### ISSUE TITLE
Duplicate header handling as error/warning ?

### ISSUE THREAD

**Issue by AvnerCohen** [UlisesGascon/check-my-headers#18]
Here is a sample output of a curl (removed the less important items) :
```
$> curl -I https://xx.xx.xxx/yyyy
HTTP/2 200
date: Tue, 21 Jul 2020 06:55:43 GMT
content-type: text/html; charset=UTF-8
content-length: 1918
vary: Accept-Encoding
content-security-policy: xxxxxx
cache-control: no-store
expires: 0
x-frame-options: SAMEORIGIN
x-xss-protection: 1; mode=block
strict-transport-security: max-age=31536000; includeSubdomains
x-content-type-options: nosniff
cache-control: no-store
```
Notice in this example the duplication of `cache-control`.

As per the RFC - https://tools.ietf.org/html/rfc7230#section-3.2.2

``` 
   A sender MUST NOT generate multiple header fields with the same field
   name in a message unless either the entire field value for that
   header field is defined as a comma-separated list [i.e., #(values)]
   or the header field is a well-known exception (as noted below).
```

I couldn't find anything specific that cover such requirement in the OWASP recommendations, but I suspect this could be just a miss out.

A risk here is the fact that a different tooling and different web servers just behaves differently, anywhere from first header takes precedence to random and undefined behavior, to merge that will cause an invalid behavior etc.

@UlisesGascon Any thoughts on that ?

**Comment by lirantal** [comment]
@AvnerCohen I'm not sure what exactly you're asking :)

**Comment by AvnerCohen** [comment]
Great, Looks like I did a good job here.. 😆 

`check-my-headers` is designed to check http headers for best practices.

One such best practice not covered here is the case of header duplication.

In the first original curl snipper curl I posted, you can see `cache-control: no-store` is fully duplicated.

This could lead to all sorts of weird and unexpected behaviours.

A simple recreation:

server.js
```nodejs
const http = require('http');

const hostname = '127.0.0.1';
const port = 3002;

const server = http.createServer((req, res) => {
  res.statusCode = 200;
  res.setHeader('Content-Type', 'text/plain');
  res.setHeader('cache-control', 'no-cache');
  res.end('Hello World');
});

server.listen(port, hostname, () => {
  console.log(`Server running at http://${hostname}:${port}/`);
});
```

Caddyfile:
```
:3001

header cache-control 'no-store'
reverse_proxy localhost:3002
```

The resulting curl headers are:
```
HTTP/1.1 200 OK
Cache-Control: 'no-store'
Cache-Control: no-cache
Content-Length: 11
Content-Type: text/plain
Date: Sun, 26 Jul 2020 10:45:50 GMT
Server: Caddy

Hello World
```

See the cache control duplication.

**Comment by lirantal** [comment]
Oh I see now. You did well explaining, probably just me :-)
So you expect check-my-headers to show duplicate headers as an issue? Reasonable and very well aligned with the RFC as you noted indeed.

Are you up for a PR?

**Comment by AvnerCohen** [comment]
@lirantal I am actually not quite sure this can be solved in the current core implementation here.

Considering this snippet:
```
const request = require('request')
request
  .get('http://127.0.0.1:3001/')
  .on('response', function(response) {
    for (let header in response.headers)
        console.log(`${header} - ${response.headers[header]}`)
  })
```

`response.headers` is an object, with headers as keys, e.g. not the original array of headers that includes the duplication.

The thing is, this is evident also in the low level http/https lib:

```
const http = require("http");
const options = {
    hostname: '127.0.0.1',
    port: 3001,
    path: '/',
    method: 'GET',
    headers: {}
};

const req = http.request(options, (res) => {
    for (let header in res.headers)
        console.log(`${header} - ${res.headers[header]}`)
});
req.end();
```

Maybe there is a further level we can drop down to really get the original stream, but not sure about that.

**Comment by AvnerCohen** [comment]
A possible alternative is, rather than using node's HTTP (and the higher level request >> request-promise) is to use a third party that does curl native binding, such as https://github.com/JCMais/node-libcurl

**Comment by lirantal** [comment]
I think the native and other libs probably conform to the spec and already give you as the result the "safe" header and not the repeated one but that needs validation, just my assumption. See the Node.js API spec on duplicate headers [here](https://nodejs.org/api/http.html#http_message_headers). Either way, doesn't help us to point out the problem, as you noted indeed.

Agree with this requiring a replace of the HTTP lib so we can catch these things.
I think it makes sense overall. Can't speak for @UlisesGascon, but I have a feeling he wouldn't mind which lib is used exactly.

**Comment by UlisesGascon** [comment]
Yeah! Seems like a great idea @AvnerCohen. I agree with @lirantal, let's find a great library that can provide the raw headers so we can alert about duplications 👍

**Comment by AvnerCohen** [comment]
I came across this while trying out a very specific use case.

What I really wanted to do is instead of checking specific rules, I wanted to see headers are as expected. So a very simple string matching, but of course with some logics.
Thinking about it more as an expectations liv rather than best practices verification.
I ended up quickly creating this on top of a curl impl - https://github.com/AvnerCohen/http-headers-verifier
I totally think this can be improved here, but I don't think I'll ever get to it, so happy to close this issue from my pov.
